src/util/cow.h

```diff
@@ -0,0 +1,235 @@
+/*******************************************************************\
+
+Module: Copy on write class
+
+Author: Reuben Thomas, reuben.thomas@diffblue.com
+
+\*******************************************************************/
+
+#ifndef CPROVER_UTIL_COW_H
+#define CPROVER_UTIL_COW_H
+
+#include "util/small_shared_ptr.h"
+#include "util/invariant.h"
+
+#include <limits>
+
+template <typename T>
+class copy_on_writet final
```

Comment:
Could you please add a comment briefly describing the idea and capabilities of that class?